cpp/src/arrow/array/builder_dict.h

```diff
@@ -409,6 +409,10 @@ class DictionaryBuilder : public internal::DictionaryBuilderBase<AdaptiveIntBuil
   using BASE = internal::DictionaryBuilderBase<AdaptiveIntBuilder, T>;
   using BASE::BASE;
 
+  Status ExpandIndexByteWidth(uint8_t new_index_byte_width) {
+    return BASE::indices_builder_.ExpandIntSize(new_index_byte_width);
```

Comment:
Wouldn't it be better to add a new constructor signature, such as:
```c++
DictionaryBuilderBase(const std::shared_ptr<DataType>& start_index_type,
                      const std::shared_ptr<DataType>& value_type,
                      MemoryPool* pool = default_memory_pool())
```
?

Reply:
I see. I try to rewrite this way.

Reply:
@pitrou I've done to do this.  Could you please review this again?